### PR TITLE
CMake: Use -Wold-style-cast

### PR DESCRIPTION
**Pull request by spycrab**
GCC and Clang will now show a warning every time a C-Style cast is used.

### CODE REVIEW

**Comment by ligfx** [comment]
Please don't enable this until the preexisting old-style casts are cleaned up.

**Comment by spycrab** [comment]
Now our warnings shouldn't be present in *most* Externals

**Comment by spycrab** [comment]
@ligfx Agreed. Clean up in separate PR or here?

**Comment by lioncash** [comment]
I'm not sure marking all of externals as system libraries is the proper way to go about fixing this. The warning flag additions should likely just be moved into the CMakeLists file in Source/. `check_and_add_flag` internally uses [add_compile_options](https://cmake.org/cmake/help/git-master/command/add_compile_options.html) which adds flags to all targets in the current directory and its subdirectories. Since it's in the top-level CMakeLists file, it's also including the targets in externals.

@Orphis any thoughts on this.

**Comment by spycrab** [comment]
@lioncash For some reason I *still* get -Wold-style-cast warnings when compiling Externals e.g. GLSLang, despite ``check_and_add_flag`` being in the Source directory...

**Comment by spycrab** [comment]
After disabling Warnings for Wx, log size is down from 25 MB to 40 KB.

**Comment by Orphis** [comment]
I agree with marking externals with SYSTEM. That's the proper way to resolve most warnings.
But `-w` isn't the proper way to disable warnings on Windows, there you want `/W0` instead. Otherwise, you'll have tons of other warnings `D9025: overriding option /W2 with /w`

**Comment by spycrab** [comment]
@Orphis 
Does this look sane to you?
```cmake
if (not MSVC)
  add_definitions(-w)
else()
  add_definitions(/W0)
endif()
```

I think that would cover it.

**Comment by Orphis** [comment]
```
add_compile_options(
  $<$<C_COMPILER_ID:MSVC>:/W0>
  $<$<NOT:$<C_COMPILER_ID:MSVC>>:-w>
)
```

Too bad we don't have CMake 3.8 or we could just use `$<IF:$<C_COMPILER_ID:MSVC>,/W0,-w>`

**Comment by ligfx** [comment]
@spycrab I don't mind whether it's this PR or another one, as long as it's done before this warning is enabled

File: Externals/wxWidgets3/CMakeLists.txt [hunk 1]
```diff
@@ -895,4 +895,10 @@ else()
         message(FATAL_ERROR "wxWidgets in Externals is not compatible with your platform")
 endif()
 
+if (MSVC)
+	add_definitions(/W0)
```
Comment (Orphis) [inline Externals/wxWidgets3/CMakeLists.txt]: If you don't use what I suggested, please at least use `add_compile_options()` or even `target_compile_options(lib PRIVATE ...)`

**Comment by spycrab** [comment]
Will reopen this when it's appropriate.

**Pull request closed without merging**